# HG changeset patch
# User Jaroslav Hajek <highegg@gmail.com>
# Date 1265625127 -3600
#      Mon Feb 08 11:32:07 2010 +0100
# Node ID db613bccd9922423bff5e787d7048ffb0ce9e3aa
# Parent  3a8c13b71612471c5db8f03e98f9ae98f6094c84
take advantage of new sort optimization in accumarray

diff --git a/scripts/ChangeLog b/scripts/ChangeLog
--- a/scripts/ChangeLog
+++ b/scripts/ChangeLog
@@ -1,8 +1,13 @@
+2010-02-08  Jaroslav Hajek  <highegg@gmail.com>
+
+	* general/accumarray.m: Add new test that also forces index cache
+	generation for faster sorting.
+
 2010-02-08  Jaroslav Hajek  <highegg@gmail.com>
 
 	* general/accumarray.m: 1 more small fix.
 
 2010-02-06  Jaroslav Hajek  <highegg@gmail.com>
 
 	* general/accumarray.m: Small fixes.
 
diff --git a/scripts/general/accumarray.m b/scripts/general/accumarray.m
--- a/scripts/general/accumarray.m
+++ b/scripts/general/accumarray.m
@@ -142,19 +142,21 @@ function A = accumarray (subs, val, sz =
       elseif (ndims != length (sz))
         error ("accumarray: dimensions mismatch")
       endif
 
       ## Convert multidimensional subscripts.
       if (ismatrix (subs))
         subs = num2cell (subs, 1);
       endif
-      subs = sub2ind (sz, subs{:});
+      subs = sub2ind (sz, subs{:}); # creates index cache
     elseif (! isempty (sz) && length (sz) < 2)
       error ("accumarray: needs at least 2 dimensions");
+    elseif (! isindex (subs)) # creates index cache
+      error ("accumarray: indices must be positive integers");
     endif
 
 
     ## Some built-in reductions handled efficiently.
 
     if (isempty (func) || func == @sum)
       ## Fast summation.
       if (isempty (sz))
